// react-native-oreopsnew/components/OreMap/index.js
// IOP Code for Map Control is CNT_MC_954
import React, { Component } from 'react';  
import { StyleSheet} from 'react-native';  
import MapView from 'react-native-maps';  
import { Marker } from 'react-native-maps';  

// Map Control Template
export default class OreMap extends Component {  
  render() {  
    return (  
       <MapView mapType={this.props.mapType} 
          style={[styles.mapStyle,this.props.style]}  
          region={{  
            latitude: this.props.latitude,
            longitude: this.props.longitude,
            latitudeDelta: 0.0922,  
            longitudeDelta: 0.0421,  
            
          }}  >  
       <Marker pinColor={this.props.markerColor} 
            coordinate={{ latitude: this.props.latitude, longitude: this.props.longitude }}  
            title={this.props.markertitle}  
            description={this.props.markerdescription}  
        />  
       </MapView>  
     );  
  }  
}  
 
// Default CSS Property for Map Control
const styles = StyleSheet.create({  
   mapStyle: {  
    alignItems: 'center',  
    justifyContent: 'flex-end', 
   },  
});
