// react-native-oreopscore/components/OreOPS/connectors/Ore_SMS.js
import axios from 'axios';
import connections from '../../../../../connector/Ore_Connections';

// Method for returning object from response data
const OreResponse = (status, statuscode, data, error) => {
    return new Promise(function (resolve) {
        resolve({ status: status, statuscode: statuscode, data: data, error: error });
    });
}

// Method for Sending Sms to device
const SendSMS = async (conname, phone, message) => {
    try {
        let config = connections.GetConnectorConfig(conname);
        if (!config) {
            return OreResponse(false, 206, [], conname + ": configuration not exist");
        }
        if (!phone) {
            return OreResponse(false, 206, [], "Phone Required Ex: 123456789");
        }
        if (!message) {
            return OreResponse(false, 206, [], "Message Required");
        }

        let TYPE = 'POST';
        let URL = connections.OreCommURL + 'SendSMS';
        let PARAMS = {
            pnConnType: 170,
            Country: Number(config.country),
            To: phone.toString(),
            Message: message,
            Sender: config.sender,
            Route: Number(config.route),
            Credential: JSON.stringify({ Authkey: config.authkey, Secret: '' })
        }
        let HEADERS = {};
        if (!TYPE) {
            return OreResponse(false, 206, [], "TYPE Required Ex: POST");
        }
        if (!URL) {
            return OreResponse(false, 206, [], "URL Required");
        }

        const response = await axios.post(URL, PARAMS, { headers: HEADERS });
        let result = response.data;
        result = result[0];
        if (result.Status) {
            return OreResponse(true, result.Status, result.Data, []);
        }
        else {
            return OreResponse(false, result.Status, [], result.ErrorLog);
        }
    }
    catch (err) {
        return OreResponse(false, 404, [], err.toString());
    }
}

export default { SendSMS }
