// react-native-oreopsnew/components/OreStyle/OreStyleDark.js
import { StyleSheet} from 'react-native';
import Oreframework from 'react-native-oreopscore/components/OreOPS/general/Ore_Framework';

const OreDarkGlobal = {
  //Color
 color:{
  pageBackgroundColor:"#252F4A",
  primaryColor: "#41B8B4",
  primaryLightColor:"#C9F7f6",
  primaryColorDrawer : "#252F4A",
  secondaryColor1:"#41B8B4",
  secondaryColor2:"#41B8B4",
  secondaryColor3:"#41B8B4",
  activeColor:"#FFFFFF",
  activeBackgroundColor:"#303E62",
  inactiveColor:"#CCCCCC",
  borderColor:"#FFFFFF",
  labelFontColor:"#FFFFFF",
  formControlFontColor:"#FFFFFF",
  actionControlFontColor:"#FFFFFF",
  
},

//Font
font:{
fontFamily:"normal",
headingFontSize: Oreframework.normalize(14),
titleFontSize: Oreframework.normalize(16),
subtitleFontSize: Oreframework.normalize(14),
bodyFontSize: Oreframework.normalize(12),
labelFontSize: Oreframework.normalize(11),
formFontSize: Oreframework.normalize(13),
drawerTextSize: Oreframework.normalize(11),

},

//Layout
layout:{
 formControlWidth:"90%",
 formControlHeight:45,
 pagePadding:0,
},

//Opacity
opacity:{
 activeOpacity: 0.5,
 disabledOpacity: 0.5
},

//Border
border:{
 borderType:"square",
 borderStyle:"solid",
 formControlBorderRadius: 0,
 formControlRoundedBorderRadius: 50,
 borderWidth:1,
 },

//Icon
icon:{
  small: 25,
  Medium: 30,
  Large: 35,
  ExtraLarge: 40
},

//Text Property
textProps:{
  selectable: false
},

//Icon Property
iconProps:{
  iconType : "regular",
  iconSize : 25,
  iconColor : "#FFFFFF",
  iconColorRounded : "#FFFFFF"
},

 //Image Property
 imageProps:{
  resizeMode:"stretch"
},
//qrCode Property
qrCodeProps: {
   logoBorderRadius: 0,
   size: 100,
},

// Html Property
htmlProps: {
  originWhitelist: ['*'],
  horizontalindicator: true,
  verticalindicator: true,
},

// Radio Property
radioButtonProps: {
  flexDirection: 'row'
},

// Checkbox Property
checkBoxProps: {
  flexDirection: 'row',

},

//Date Picker Property
datePickerProps: {
  mode:'date',
  format:'DD-MM-YYYY',
  iconstyle:'right',
  iconsize:30,
  iconname:'calendar',
  iconcolor:'#FFFFFF'
},

//File Picker Properties
filePickerProps: {
  cameraPickerAnimDuration : 250,
  cameraPickerHight: 120,
  placeholderTextColor : '#CCCCCC',
  closeIconColor : '#FFFFFF',
  closeIconName : 'close',
  closeIconSize : 15,
  cameraPickerLabel : 'Choose option',
  cameraPickerCameraTxt : 'Take Photo',
  cameraPickerGallaryTxt : 'Choose from gallary'
},
//button strip props
bottomTabStrip: {
  borderTopWidth: 1,
  borderRightWidth: 0
},
//Listview Props 
listViewProps:{
  listCartHeadingColor:"#FAFAFA",
  listCartContentColor:"#E1E1E1",
  listCartPriceColor:"#FFFFFF",
  listCartOfferColor:"#FF0000",
  listCartBorderColor:"#505461",
  listStyle4backgroundColor:"#F1F1F1",
  listStyle4TextColor:"#333333",
  listStyle4headingColor:"#FFFFFF",
  listHeadingSize:Oreframework.normalize(11),
  listContentSize:Oreframework.normalize(9),
  listSmallSize:Oreframework.normalize(8),
  listStyle4TextSize:Oreframework.normalize(10),
  listStyle4HeadingSize:Oreframework.normalize(12),
}

}

 

const OreStyleDark = StyleSheet.create({
 // Page Style   
 PageStyle : { 
  flex: 1, 
  width: "100%", 
  height: "100%", 
  backgroundColor: OreDarkGlobal.color.pageBackgroundColor,
},

// Text Control Style
OreText: {
width:  OreDarkGlobal.layout.formControlWidth,
height : OreDarkGlobal.layout.formControlHeight,
color:  OreDarkGlobal.color.formControlFontColor,
backgroundColor: "transparent",
textAlign:"auto",
textDecorationLine:"none",
textTransform:"none",
fontSize: OreDarkGlobal.font.formFontSize,
//fontFamily: OreDarkGlobal.font.fontFamily === 'normal' ? Platform.OS === "ios" ? 'San Francisco' : 'normal' : OreDarkGlobal.font.fontFamily,
borderWidth: 0,
borderColor: OreDarkGlobal.color.borderColor,
borderRadius: OreDarkGlobal.border.borderType == "rounded" ? OreDarkGlobal.border.formControlRoundedBorderRadius : OreDarkGlobal.border.formControlBorderRadius,
borderStyle: OreDarkGlobal.border.borderStyle
},

 // Icon Style2
 OreIconRounded:{
  backgroundColor: OreDarkGlobal.color.primaryColor,
  borderRadius: OreDarkGlobal.border.formControlRoundedBorderRadius,
  width: 50,
  height: 50,
  justifyContent: "center",
  alignItems:"center",
 },
 OreIcon:{
   justifyContent: "center",
   alignItems:"center",
   borderWidth: 0,
   borderColor: OreDarkGlobal.color.borderColor,
   borderRadius: OreDarkGlobal.border.borderType == "rounded" ? OreDarkGlobal.border.formControlRoundedBorderRadius : OreDarkGlobal.border.formControlBorderRadius,
   borderStyle: OreDarkGlobal.border.borderStyle
 },
 

//Image Style
OreImageViewStyle:{
  width:  OreDarkGlobal.layout.formControlWidth,
  height : OreDarkGlobal.layout.formControlHeight,
  backgroundColor: "transparent",
  borderWidth: 0,
  borderColor: OreDarkGlobal.color.borderColor,
  borderRadius: OreDarkGlobal.border.borderType === "rounded" ? OreDarkGlobal.border.formControlRoundedBorderRadius : OreDarkGlobal.border.formControlBorderRadius,
  borderStyle: OreDarkGlobal.border.borderStyle
},
OreImageStyle:{
   alignSelf:"stretch",
   width:"auto",
   flex:1,
   borderRadius: OreDarkGlobal.border.borderType === "rounded" ? OreDarkGlobal.border.formControlRoundedBorderRadius : OreDarkGlobal.border.formControlBorderRadius,
},

//Button Style 
OreButtonStyle: {
  backgroundColor: OreDarkGlobal.color.primaryColor,
  width:  OreDarkGlobal.layout.formControlWidth,
  height : OreDarkGlobal.layout.formControlHeight,
  justifyContent: 'center',
  borderWidth: 0,
  borderColor: OreDarkGlobal.color.borderColor,
  borderRadius: OreDarkGlobal.border.borderType === "rounded" ? OreDarkGlobal.border.formControlRoundedBorderRadius : OreDarkGlobal.border.formControlBorderRadius,
  borderStyle: OreDarkGlobal.border.borderStyle
},
OreButtonTextStyle: {
  color: OreDarkGlobal.color.actionControlFontColor,
  textAlign: 'center',
  fontSize: OreDarkGlobal.font.formFontSize,
  //fontFamily: OreDarkGlobal.font.fontFamily === 'normal' ? Platform.OS === "ios" ? 'San Francisco' : 'normal' : OreDarkGlobal.font.fontFamily,
},

//Switch Style
OreSwitch:{
transform: [{ scaleX: 1.2 }, { scaleY: 1.2 }],
},
//Qr Code Style
OreQRCodeStyle: {
  width:  100,
  height: 100, 
 },
// Radio Button view style
OreRadioButtonRoot: {
  flexDirection: 'row',
  marginLeft: -6,
  marginTop: -2,
  marginRight : 7
},

// Radio Button Text Style
OreRadioButtonTextstyle: {
  color: OreDarkGlobal.color.formControlFontColor,
  fontSize: OreDarkGlobal.font.formFontSize,
  alignSelf: 'center',
  fontFamily: OreDarkGlobal.font.fontFamily,
  fontWeight: OreDarkGlobal.font.fontWeight,
},

// checkbox view style
OreCheckboxRoot: {
  flexDirection: 'row',
  alignItems: 'flex-start',
  alignSelf: 'flex-start',
  justifyContent: 'flex-start'
},

// Checkbox Text  style
OreCheckboxText: {
  fontSize: OreDarkGlobal.font.formFontSize,
  color: OreDarkGlobal.color.formControlFontColor,
  fontFamily: OreDarkGlobal.font.fontFamily,
  alignSelf: 'center',
},
//Textbox view style
OreTextBoxViewStyle: {
  width: '100%'
},

//Textbox label style
OreTextBoxLabelStyle: {
  textAlignVertical: 'bottom',
  fontSize: OreDarkGlobal.font.labelFontSize,
  color: OreDarkGlobal.color.labelFontColor,
  fontFamily: OreDarkGlobal.font.fontFamily,
},

//Textbox style
OreTextBox:
{
  color: OreDarkGlobal.color.formControlFontColor,
  fontSize: OreDarkGlobal.font.formFontSize,
  fontFamily: OreDarkGlobal.font.fontFamily,
  borderColor: OreDarkGlobal.color.borderColor,
  borderWidth: OreDarkGlobal.border.borderType !== "flat" ? OreDarkGlobal.border.borderWidth : 0,
  borderBottomWidth : OreDarkGlobal.border.borderType === "flat" ? OreDarkGlobal.border.borderWidth : null,
  borderStyle: OreDarkGlobal.border.borderStyle,
  borderRadius : OreDarkGlobal.border.borderType === "rounded" ? OreDarkGlobal.border.formControlRoundedBorderRadius : OreDarkGlobal.border.formControlBorderRadius,
  paddingLeft: OreDarkGlobal.border.borderType !== "flat" ? 10 : 0, 
  paddingRight: OreDarkGlobal.border.borderType !== "flat" ? 10 : 0,
  marginBottom: OreDarkGlobal.border.borderType !== "flat" ? 5 : 0,
},

//Picker label style
OrePickerLabelStyle: {
  textAlignVertical: 'bottom',
  fontSize: OreDarkGlobal.font.labelFontSize,
  color: OreDarkGlobal.color.labelFontColor,
  fontFamily: OreDarkGlobal.font.fontFamily,
},

//Picker style
OrePicker:
{
  color: OreDarkGlobal.color.formControlFontColor,
  fontSize: OreDarkGlobal.font.formFontSize,
  fontFamily: OreDarkGlobal.font.fontFamily,
  borderColor: OreDarkGlobal.color.borderColor,
  borderWidth: OreDarkGlobal.border.borderType !== "flat" ? OreDarkGlobal.border.borderWidth : 0,
  borderBottomWidth : OreDarkGlobal.border.borderType === "flat" ? OreDarkGlobal.border.borderWidth : null,
  borderStyle: OreDarkGlobal.border.borderStyle,
  borderRadius : OreDarkGlobal.border.borderType === "rounded" ? OreDarkGlobal.border.formControlRoundedBorderRadius : OreDarkGlobal.border.formControlBorderRadius,
  paddingLeft: OreDarkGlobal.border.borderType !== "flat" ? 10 : 0, 
  paddingRight: OreDarkGlobal.border.borderType !== "flat" ? 10 : 0,
  marginBottom: OreDarkGlobal.border.borderType !== "flat" ? 5 : 0,
},

//Picker Text Style
OrePickerTextStyle:{
  color: OreDarkGlobal.color.formControlFontColor,
},

//Date picker label style
OreDatePickerTitleStyle: {
  fontSize: OreDarkGlobal.font.labelFontSize,
  fontFamily: OreDarkGlobal.font.fontFamily,
  color: OreDarkGlobal.color.labelFontColor,
  fontWeight: "normal",
  textAlign: "left"
},

//Date picker date style
OreDatePickerDateStyle: {
  fontSize: OreDarkGlobal.font.formFontSize,
  fontFamily: OreDarkGlobal.font.fontFamily,
  fontWeight: "normal",
  color: OreDarkGlobal.color.formControlFontColor
},

//Date picker date style
OreDatePickerDateView: {
  borderWidth: OreDarkGlobal.border.borderWidth,
  borderColor: OreDarkGlobal.color.borderColor,
  marginLeft:36
},

//File Picker style1
OreFilePickerSubViewStyle_1: {
  borderWidth: OreDarkGlobal.border.borderWidth,
  borderStyle: "solid",
  borderColor: OreDarkGlobal.color.borderColor,
  borderBottomWidth : OreDarkGlobal.border.borderWidth,
  borderRadius : 0,
},

OreFilePickerButtonStyle_1: {
  backgroundColor: OreDarkGlobal.color.primaryColor,
},

OreFilePickerLabelStyle_1: {
  fontSize: Oreframework.normalize(11),
  fontFamily: OreDarkGlobal.font.fontFamily,
},

OreFilePickerTextStyle_1: {
  fontSize: Oreframework.normalize(OreDarkGlobal.font.bodyFontSize),
  fontFamily: OreDarkGlobal.font.fontFamily,
  fontWeight: "normal",
  color: OreDarkGlobal.color.actionControlFontColor
},

//File Picker style2
OreFilePickerSubViewStyle_2: {
  borderWidth: OreDarkGlobal.border.borderWidth,
  borderStyle: "solid",
  borderColor: OreDarkGlobal.color.borderColor,
  borderBottomWidth : OreDarkGlobal.border.borderWidth,
  borderRadius : 0,
},

OreFilePickerButtonStyle_2: {
  borderStyle: "solid",
  backgroundColor: OreDarkGlobal.color.primaryColor,
},

OreFilePickerLabelStyle_2: {
  fontSize: Oreframework.normalize(11),
  fontFamily: OreDarkGlobal.font.fontFamily,
},

//File Picker style3

OreFilePickerButtonStyle_3: {
  borderStyle: "solid",
  borderRadius: 30,
  backgroundColor: OreDarkGlobal.color.primaryColor
},

//File Picker style4
OreFilePickerImageStyle_4: {
  width: 100,
  height: 100,
  borderRadius: 300
},

OreFilePickerIconStyle_4: {
  right: "20%",
  padding: 5,
  backgroundColor: '#f7f5f5',
  borderRadius: 50
},

OreFilePickerCloseIconStyle: {
  backgroundColor: OreDarkGlobal.color.borderColor,
},

//File picker camera selection dialog
CameraPickerRootStyle: {
  backgroundColor : OreDarkGlobal.color.pageBackgroundColor
},

CameraPickerLabelStyle: {
  fontWeight: 'bold',  
  fontSize: 18, 
  fontFamily: OreDarkGlobal.font.fontFamily,
  textAlign: 'left', 
  width: '100%'
},

CameraPickerTextStyle: {
  color: OreDarkGlobal.color.formControlFontColor, 
  fontSize: 16, 
  fontFamily: OreDarkGlobal.font.fontFamily,
  marginTop: 8, 
  width: '100%', 
  textAlign: 'left'
},

// Button Strip Control
bottomButtonStripStyle: {
  fontSize : OreDarkGlobal.font.bodyFontSize,
  fontFamily : OreDarkGlobal.font.fontFamily,
  color : OreDarkGlobal.color.actionControlFontColor
}


});
module.exports = {OreStyleDark,OreDarkGlobal};
