// react-native-oreopsnew/components/OreStyle/OreStyleLight.js
import { StyleSheet, Platform} from 'react-native';
import Oreframework from 'react-native-oreopscore/components/OreOPS/general/Ore_Framework';

const OreLightGlobal = {
  //Color
  color:{
    pageBackgroundColor:"#FFFFFF",
    primaryColor: "#3F51B5",
    primaryLightColor:"#B1B8DD",
    primaryColorDrawer : "#3F51B5",
    secondaryColor1:"#3F51B5",
    secondaryColor2:"#3F51B5",
    secondaryColor3:"#3F51B5",
    activeColor:"#FFFFFF",
    activeBackgroundColor:"#5065d1",
    inactiveColor:"#CCCCCC",
    borderColor:"#CCCCCC",
    labelFontColor:"#888888",
    formControlFontColor:"#000000",
    actionControlFontColor:"#FFFFFF",
    
  },
  
 //Font
 font:{
  fontFamily:"normal",
  headingFontSize: Oreframework.normalize(14),
  titleFontSize: Oreframework.normalize(16),
  subtitleFontSize: Oreframework.normalize(14),
  bodyFontSize: Oreframework.normalize(12),
  labelFontSize: Oreframework.normalize(11),
  formFontSize: Oreframework.normalize(13),
  drawerTextSize: Oreframework.normalize(11),
 
},

  //Layout
 layout:{
   formControlWidth:"90%",
   formControlHeight:45,
   pagePadding:0,
  },

//Opacity
  opacity:{
   activeOpacity: 0.5,
   disabledOpacity: 0.5
  },

//Border
  border:{
   borderType:"square",
   borderStyle:"solid",
   formControlBorderRadius: 0,
   formControlRoundedBorderRadius: 50,
   borderWidth: 1,
   },

  //Icon
  icon:{
    small: 25,
    Medium: 30,
    Large: 35,
    ExtraLarge: 40
  },

  //Text Property
  textProps:{
    selectable: false
  },

  //Icon Property
  iconProps:{
    iconType : "regular",
    iconSize : 25,
    iconColor : "#888888",
    iconColorRounded : "#FFFFFF"
 },

 //Image Property
  imageProps:{
     resizeMode:"stretch"
  },

  //qrCode Property
qrCodeProps: {
  logoBorderRadius: 0,
  size: 100,
},

// Html Property
htmlProps: {
  originWhitelist: ['*'],
  horizontalindicator: true,
  verticalindicator: true,
},

// Radio Property
radioButtonProps: {
  flexDirection: 'row'
},

// Checkbox Property
checkBoxProps: {
  flexDirection: 'row',
},

//Date Picker Property
datePickerProps: {
  mode:'date',
  format:'DD-MM-YYYY',
  iconstyle:'right',
  iconsize:30,
  iconname:'calendar',
  iconcolor:'#777777'
},

//File Picker Properties
filePickerProps: {
  cameraPickerAnimDuration : 250,
  cameraPickerHight: 120,
  placeholderTextColor : '#CCCCCC',
  closeIconColor : '#FFFFFF',
  closeIconName : 'close',
  closeIconSize : 15,
  cameraPickerLabel : 'Choose option',
  cameraPickerCameraTxt : 'Take Photo',
  cameraPickerGallaryTxt : 'Choose from gallary'
},

//button strip props
bottomTabStrip: {
  borderTopWidth: 1,
  borderRightWidth: 0
},

//Listview Props 
listViewProps:{
  listCartHeadingColor:"#000000",
  listCartContentColor:"#888888",
  listCartPriceColor:"#333333",
  listCartOfferColor:"#FF0000",
  listCartBorderColor:"#CCCCCC",
  listStyle4backgroundColor:"#F1F1F1",
  listStyle4TextColor:"#333333",
  listStyle4headingColor:"#333333",
  listHeadingSize:Oreframework.normalize(11),
  listContentSize:Oreframework.normalize(9),
  listSmallSize:Oreframework.normalize(8),
  listStyle4TextSize:Oreframework.normalize(10),
  listStyle4HeadingSize:Oreframework.normalize(12),
}
}

const OreStyleLight = StyleSheet.create({
  // Page Style   
  PageStyle : { 
      flex: 1, 
      width: "100%", 
      height: "100%", 
      backgroundColor: OreLightGlobal.color.pageBackgroundColor,
  },
  
  // Text Control Style
OreText: {
  width:  OreLightGlobal.layout.formControlWidth,
  height : OreLightGlobal.layout.formControlHeight,
  color:  OreLightGlobal.color.formControlFontColor,
  backgroundColor: "transparent",
  textAlign:"auto",
  textDecorationLine:"none",
  textTransform:"none",
  fontSize: OreLightGlobal.font.formFontSize,
 // fontFamily: OreLightGlobal.font.fontFamily === 'normal' ? Platform.OS === "ios" ? 'San Francisco' : 'normal' : OreLightGlobal.font.fontFamily,
  borderWidth: 0,
  borderColor: OreLightGlobal.color.borderColor,
  borderRadius: OreLightGlobal.border.borderType === "rounded" ? OreLightGlobal.border.formControlRoundedBorderRadius : OreLightGlobal.border.formControlBorderRadius,
  borderStyle: OreLightGlobal.border.borderStyle
  },

  // Icon Style2
  OreIconRounded:{
   backgroundColor: OreLightGlobal.color.primaryColor,
   borderRadius: OreLightGlobal.border.formControlRoundedBorderRadius,
   width: 50,
   height: 50,
   justifyContent: "center",
   alignItems:"center",
  },
  OreIcon:{
    justifyContent: "center",
    alignItems:"center",
    borderWidth: 0,
    borderColor: OreLightGlobal.color.borderColor,
    borderRadius: OreLightGlobal.border.borderType === "rounded" ? OreLightGlobal.border.formControlRoundedBorderRadius : OreLightGlobal.border.formControlBorderRadius,
    borderStyle: OreLightGlobal.border.borderStyle
   },
  
  //Image Style
  OreImageViewStyle:{
   width:  OreLightGlobal.layout.formControlWidth,
   height : OreLightGlobal.layout.formControlHeight,
   backgroundColor: "transparent",
   borderWidth: 0,
   borderColor: OreLightGlobal.color.borderColor,
   borderRadius: OreLightGlobal.border.borderType === "rounded" ? OreLightGlobal.border.formControlRoundedBorderRadius : OreLightGlobal.border.formControlBorderRadius,
   borderStyle: OreLightGlobal.border.borderStyle
  },
  OreImageStyle:{
    alignSelf:"stretch",
    width:"auto",
    flex:1,
    borderRadius: OreLightGlobal.border.borderType === "rounded" ? OreLightGlobal.border.formControlRoundedBorderRadius : OreLightGlobal.border.formControlBorderRadius,
  },
  
  
 //Button Style 
OreButtonStyle: {
  backgroundColor: OreLightGlobal.color.primaryColor,
  width:  OreLightGlobal.layout.formControlWidth,
  height : OreLightGlobal.layout.formControlHeight,
  justifyContent: 'center',
  borderWidth: 0,
  borderColor: OreLightGlobal.color.borderColor,
  borderRadius: OreLightGlobal.border.borderType === "rounded" ? OreLightGlobal.border.formControlRoundedBorderRadius : OreLightGlobal.border.formControlBorderRadius,
  borderStyle: OreLightGlobal.border.borderStyle
},
OreButtonTextStyle: {
  color: OreLightGlobal.color.actionControlFontColor,
  textAlign: 'center',
  fontSize: OreLightGlobal.font.formFontSize,
  //fontFamily: OreLightGlobal.font.fontFamily === 'normal' ? Platform.OS === "ios" ? 'San Francisco' : 'normal' : OreLightGlobal.font.fontFamily,
},

  
  //Switch Style
  OreSwitch:{
    transform: [{ scaleX: 1.2 }, { scaleY: 1.2 }],
  },

//Qr Code Style
  OreQRCodeStyle: {
    width:  100,
    height: 100, 
   },

   // Radio Button view style
   OreRadioButtonRoot: {
    flexDirection: 'row',
    marginLeft: -6,
    marginTop: -2,
    marginRight : 7
  },
  
  // Radio Button Text Style
  OreRadioButtonTextstyle: {
    color: OreLightGlobal.color.formControlFontColor,
    fontSize: OreLightGlobal.font.formFontSize,
    alignSelf: 'center',
    fontFamily: OreLightGlobal.font.fontFamily,
    fontWeight: OreLightGlobal.font.fontWeight,
  },

  // checkbox view style
  OreCheckboxRoot: {
    flexDirection: 'row',
    alignItems: 'flex-start',
    alignSelf: 'flex-start',
    justifyContent: 'flex-start'
  },

  // Checkbox Text  style
  OreCheckboxText: {
    fontSize: OreLightGlobal.font.formFontSize,
    fontFamily: OreLightGlobal.font.fontFamily,
    color: OreLightGlobal.color.formControlFontColor,
    alignSelf: 'center',
  },
  
  //Textbox view style
  OreTextBoxViewStyle: {
     width: '100%'
  },

  //Textbox label style
  OreTextBoxLabelStyle: {
    textAlignVertical: 'bottom',
    fontSize: OreLightGlobal.font.labelFontSize,
    fontFamily: OreLightGlobal.font.fontFamily,
    color: OreLightGlobal.color.labelFontColor,
  },

  //Textbox style
  OreTextBox:
  {
    color: OreLightGlobal.color.formControlFontColor,
    fontSize: OreLightGlobal.font.formFontSize,
    fontFamily: OreLightGlobal.font.fontFamily,
    borderColor: OreLightGlobal.color.borderColor,
    borderWidth: OreLightGlobal.border.borderType !== "flat" ? OreLightGlobal.border.borderWidth : 0,
    borderBottomWidth : OreLightGlobal.border.borderType === "flat" ? OreLightGlobal.border.borderWidth : null,
    borderStyle: OreLightGlobal.border.borderStyle,
    borderRadius : OreLightGlobal.border.borderType === "rounded" ? OreLightGlobal.border.formControlRoundedBorderRadius : OreLightGlobal.border.formControlBorderRadius,
    paddingLeft: OreLightGlobal.border.borderType !== "flat" ? 10 : 0, 
    paddingRight: OreLightGlobal.border.borderType !== "flat" ? 10 : 0,
    marginBottom: OreLightGlobal.border.borderType !== "flat" ? 5 : 0,
  },

  //Picker label style
  OrePickerLabelStyle: {
    textAlignVertical: 'bottom',
    fontSize: OreLightGlobal.font.labelFontSize,
    fontFamily: OreLightGlobal.font.fontFamily,
    color: OreLightGlobal.color.labelFontColor,
  },

  //Picker style
  OrePicker:
  {
    color: OreLightGlobal.color.formControlFontColor,
    fontSize: OreLightGlobal.font.formFontSize,
    fontFamily: OreLightGlobal.font.fontFamily,
    borderColor: OreLightGlobal.color.borderColor,
    borderWidth: OreLightGlobal.border.borderType !== "flat" ? OreLightGlobal.border.borderWidth : 0,
    borderBottomWidth : OreLightGlobal.border.borderType === "flat" ? OreLightGlobal.border.borderWidth : null,
    borderStyle: OreLightGlobal.border.borderStyle,
    borderRadius : OreLightGlobal.border.borderType === "rounded" ? OreLightGlobal.border.formControlRoundedBorderRadius : OreLightGlobal.border.formControlBorderRadius,
    paddingLeft: OreLightGlobal.border.borderType !== "flat" ? 10 : 0, 
    paddingRight: OreLightGlobal.border.borderType !== "flat" ? 10 : 0,
    marginBottom: OreLightGlobal.border.borderType !== "flat" ? 5 : 0,
  },
  
  //Picker Text Style
  OrePickerTextStyle:{
    color: OreLightGlobal.color.formControlFontColor,
  },
  
 //Date picker label style
 OreDatePickerTitleStyle: {
  fontSize: OreLightGlobal.font.labelFontSize,
  fontFamily: OreLightGlobal.font.fontFamily,
  color: OreLightGlobal.color.labelFontColor,
  fontWeight: "normal",
  textAlign: "left"
},

//Date picker date style
OreDatePickerDateStyle: {
  fontSize: OreLightGlobal.font.formFontSize,
  fontFamily: OreLightGlobal.font.fontFamily,
  fontWeight: "normal",
  color: OreLightGlobal.color.formControlFontColor
},

//Date picker date style
OreDatePickerDateView: {
  borderWidth: OreLightGlobal.border.borderWidth,
  borderColor: OreLightGlobal.color.borderColor,
  marginLeft:36
},

//File Picker style1
OreFilePickerSubViewStyle_1: {
  borderWidth: OreLightGlobal.border.borderWidth,
  borderStyle: "solid",
  borderColor: OreLightGlobal.color.borderColor,
  borderBottomWidth : OreLightGlobal.border.borderWidth,
  borderRadius : 0,
},

OreFilePickerButtonStyle_1: {
  backgroundColor: OreLightGlobal.color.primaryColor,
},

OreFilePickerLabelStyle_1: {
  fontSize: Oreframework.normalize(11),
  fontFamily: OreLightGlobal.font.fontFamily,
},

OreFilePickerTextStyle_1: {
  fontSize: Oreframework.normalize(OreLightGlobal.font.bodyFontSize),
  fontFamily: OreLightGlobal.font.fontFamily,
  fontWeight: "normal",
  color: OreLightGlobal.color.actionControlFontColor
},

//File Picker style2
OreFilePickerSubViewStyle_2: {
  borderWidth: OreLightGlobal.border.borderWidth,
  borderStyle: "solid",
  borderColor: OreLightGlobal.color.borderColor,
  borderBottomWidth : OreLightGlobal.border.borderWidth,
  borderRadius : 0,
},

OreFilePickerButtonStyle_2: {
  borderStyle: "solid",
  backgroundColor: OreLightGlobal.color.primaryColor,
},

OreFilePickerLabelStyle_2: {
  fontSize: Oreframework.normalize(11),
  fontFamily: OreLightGlobal.font.fontFamily,
},

//File Picker style3

OreFilePickerButtonStyle_3: {
  borderStyle: "solid",
  borderRadius: 30,
  backgroundColor: OreLightGlobal.color.primaryColor
},

//File Picker style4
OreFilePickerImageStyle_4: {
  width: 100,
  height: 100,
  borderRadius: 300
},

OreFilePickerIconStyle_4: {
  right: "20%",
  padding: 5,
  backgroundColor: '#f7f5f5',
  borderRadius: 50
},

OreFilePickerCloseIconStyle: {
  backgroundColor: OreLightGlobal.color.borderColor,
},

//File picker camera selection dialog
CameraPickerRootStyle: {
  backgroundColor : OreLightGlobal.color.pageBackgroundColor
},

CameraPickerLabelStyle: {
  fontWeight: 'bold',  
  fontSize: 18,
  fontFamily: OreLightGlobal.font.fontFamily, 
  textAlign: 'left', 
  width: '100%'
},

CameraPickerTextStyle: {
  color: OreLightGlobal.color.formControlFontColor, 
  fontSize: 16, 
  fontFamily: OreLightGlobal.font.fontFamily,
  marginTop: 8, 
  width: '100%', 
  textAlign: 'left'
},

// Button Strip Control 

bottomButtonStripStyle: {
  fontSize : OreLightGlobal.font.bodyFontSize,
  fontFamily : OreLightGlobal.font.fontFamily,
  color : OreLightGlobal.color.actionControlFontColor
}

  });



module.exports = {OreStyleLight,OreLightGlobal};
